extensions/wikia/Captcha/Module/FancyCaptcha.class.php

```diff
@@ -0,0 +1,316 @@
+				$char = $dirs[( $place + $j ) % count( $dirs )];
+				$return = $this->pickImageDir( "$directory/$char", $levels - 1 );
+				if ( $return ) {
+					return $return;
+				}
+			}
+			// Didn't find any images in this directory... empty?
+			return false;
+		} else {
+			return $this->pickImageFromDir( $directory );
+		}
+	}
+
+	/**
+	 * @param string $directory
+	 *
+	 * @return array|bool
+	 */
+	public function pickImageFromDir( $directory ) {
+		if ( !is_dir( $directory ) ) {
+			return false;
+		}
+		$n = mt_rand( 0, $this->countFiles( $directory ) - 1 );
+		$dir = opendir( $directory );
+
+		$count = 0;
+
+		$entry = readdir( $dir );
+		$pick = false;
+		while ( false !== $entry ) {
+			$entry = readdir( $dir );
+			if ( preg_match( '/^image_([0-9a-f]+)_([0-9a-f]+)\\.png$/', $entry, $matches ) ) {
+				$size = getimagesize( "$directory/$entry" );
+				$pick = [
+					'salt' => $matches[1],
+					'hash' => $matches[2],
+					'width' => $size[0],
+					'height' => $size[1],
+					'viewed' => false,
+				];
+				if ( $count++ == $n ) {
+					break;
+				}
+			}
+		}
+		closedir( $dir );
+		return $pick;
+	}
+
+	/**
+	 * Count the number of files in a directory.
+	 *
+	 * @param string $dirName
+	 *
+	 * @return int
+	 */
+	public function countFiles( $dirName ) {
+		$dir = opendir( $dirName );
+		$count = 0;
+		while ( false !== ( $entry = readdir( $dir ) ) ) {
+			if ( $entry != '.' && $entry != '..' ) {
+				$count++;
+			}
+		}
+		closedir( $dir );
+		return $count;
+	}
+
+	/**
+	 * @return bool
+	 *
+	 * @throws \MWException
+	 */
+	public function showImage() {
+		$wg = \F::app()->wg;
+		$error = null;
+
+		$wg->Out->disable();
+
+		$info = $this->retrieveCaptcha();
+		if ( $info ) {
+			$info['viewed'] = wfTimestamp();
+			$this->storeCaptcha( $info );
+
+			$salt = $info['salt'];
+			$hash = $info['hash'];
+			$file = $this->imagePath( $salt, $hash );
+
+			if ( file_exists( $file ) ) {
+				header( "Cache-Control: private, s-maxage=0, max-age=3600" );
+				\StreamFile::stream( $file );
+				return true;
+			} else {
+				$error = 'File ' . $file . ' does not exist';
+			}
+		} else {
+			$error = 'Info is empty';
+		}
+		wfHttpError( 404, '404 not found', 'Requested non-existing captcha image' );
+		$this->log( 'Captcha returned 404: ' . $error );
+
+		return false;
+	}
+
+	/**
+	 * @param $salt
+	 * @param $hash
+	 *
+	 * @return string
+	 */
+	public function imagePath( $salt, $hash ) {
+		$file = \F::app()->wg->CaptchaDirectory;
+		$file .= DIRECTORY_SEPARATOR;
+		for ( $i = 0; $i < self::DIRECTORY_LEVELS; $i++ ) {
+			$file .= $hash { $i } ;
+			$file .= DIRECTORY_SEPARATOR;
+		}
+		$file .= "image_{$salt}_{$hash}.png";
+		return $file;
+	}
+
+	/**
+	 * Show a message asking the user to enter a captcha on edit
+	 * The result will be treated as wiki text
+	 *
+	 * @param string $action Action being performed
+	 *
+	 * @return string
+	 */
+	public function getMessage( $action ) {
+		$name = 'fancycaptcha-' . $action;
```

Comment:
Same question about thoughts on adding the rest of the message keys to the method docs to make them easier to search for?